examples/shor/functions.py

```diff
@@ -0,0 +1,420 @@
+    """
+    for i in range(2, int(np.floor(np.log2(N))) + 1):
+        f = N**(1/i)
+        f1 = np.floor(f)
+        if f1**i == N:
+            return f1
+        f2 = np.ceil(f)
+        if f2**i == N:
+            return f2
+    return None
+
+
+def egcd(a, b):
+    """Extended Euclid's algorithm for the modular inverse calculation.
+    """
+    if a == 0:
+        return (b, 0, 1)
+    else:
+        g, y, x = egcd(b % a, a)
+        return (g, x - (b // a) * y, y)
+
+
+def mod_inv(a, N):
+    """Classical algorithm to calculate the inverse of a modulo N.
+    Args:
+        a (int): number to invert.
+        N (int): modulo.
+
+    Returns:
+        x%N (int): inverse of a modulo N.
+    """
+    g, x, y = egcd(a, N)
+    if g != 1:
+        raise ValueError('modular inverse does not exist')
+    else:
+        return x % N
+
+
+def quantum_order_finding_full(N, a):
+    """Quantum circuit that performs the order finding algorithm using a fully quantum iQFT.
+    Args:
+        N (int): number to factorize.
+        a (int): chosen number to use in the algorithm.
+
+    Returns:
+        s (float): value of the state measured by the quantum computer.
+    """
+    print('  - Performing algorithm using a fully quantum iQFT.\n')
+    # Creating the parts of the needed quantum circuit.
+    n = int(np.ceil(np.log2(N)))
+    b = [i for i in range(n+1)]
+    x = [n+1+i for i in range(n)]
+    ancilla = 2*n + 1
+    q_reg = [2*n + 2 + i for i in range(2*n)]
+    circuit = Circuit(4*n+2)
+    print(f'  - Total number of qubits used: {4*n+2}.\n')
+
+    # Building the quantum circuit
+    for i in range(len(q_reg)):
+        circuit.add(gates.H(q_reg[i]))
+    circuit.add(gates.X(x[len(x)-1]))
+    circuit.add((c_U(q, x, b, a**(2**i), N, ancilla, n) for i, q in enumerate(q_reg)))
+    circuit.add(i_qft(q_reg))
+
+    # Adding measurement gates
+    circuit.add(gates.M(*q_reg))
+    result = circuit(nshots=1)
+    s = result.frequencies(binary=False).most_common()[0][0]
+    print(f"The quantum circuit measures s = {s}.\n")
+    return s
+
+
+def quantum_order_finding_semiclassical(N, a):
+    """Quantum circuit that performs the order finding algorithm using a semiclassical iQFT.
+    Args:
+        N (int): number to factorize.
+        a (int): chosen number to use in the algorithm.
+
+    Returns:
+        s (float): value of the state measured by the quantum computer.
+    """
+    print('  - Performing algorithm using a semiclassical iQFT.\n')
+    # Creating the parts of the needed quantum circuit.
+    n = int(np.ceil(np.log2(N)))
+    b = [i for i in range(n+1)]
+    x = [n+1+i for i in range(n)]
+    ancilla = 2*n + 1
+    q_reg = 2*n + 2
+    circuit = Circuit(2*n+3)
+    print(f'  - Total number of qubits used: {2*n+3}.\n')
+    r = []
+    
+    # Building the quantum circuit
+    circuit.add(gates.H(q_reg))
+    circuit.add(gates.X(x[len(x)-1]))
+    a_i = (a**(2**(2*n - 1)))
```

Comment:
I explored a bit more why the semiclassical approach crashes for 29 qubits and the problem seems to be this line. @igres26, if you like to confirm this you can simply print something right before calculating `a_i` and something right after and try to run this for a large number (eg. `python main.py --semiclassical --N 7781`) in a machine with sufficient memory (eg. DGX).

The issue is that `a_i` is a huge number which presumably Python cannot handle. For example if we have 29 qubits semiclassical, this means that `n=13` so `2**(2*n-1)=2**25` and then you attempt to calculate `a_i = a ** (2 ** 25)` which is huge if a > 1. This number has log10(a_i) ~ 2 ** 25 log10(a) > 30 million digits so it is probably not a good idea to work with it. I cannot think of any easy trick around this. Since this number is probably a phase of some gate, is there any possibility that it could be simplified using periodicity? (usually in phases you can subtract multiples of 2pi)

Reply:
Umm... that makes sense. The thing here is that we want to perform the modular exponentiation of `a**x` with `x` up to `N**2`, that is why it goes as much as `a**(2**(2*n-1))`. I suppose that when a is close to N, this number gets too big for python to keep track of correctly. 

This is not a phase of a gate, but it is part of the modular exponentiation, so it might be the case that we can apply some modulo N before calculating it. A modulo N operation is done just before calculating the angles of the needed gates.